Mathlib/CategoryTheory/MorphismProperty/Composition.lean

```diff
@@ -198,6 +198,89 @@ instance inf {P Q : MorphismProperty C} [P.IsMultiplicative] [Q.IsMultiplicative
+
+/-- A variant of multiplicativeClosure in which compositions are taken on the left rather than
+on the right. It is not intended to be used directly, and one should rather access this via
+`multiplicativeClosure.eq_prime` in cases where the inductive principle of this variant is
+needed. -/
+inductive multiplicativeClosure' (W : MorphismProperty C) : MorphismProperty C
+  | of {x y : C} (f : x ⟶ y) (hf : W f) : multiplicativeClosure' W f
+  | id (x : C) : multiplicativeClosure' W (𝟙 x)
+  | of_comp {x y z : C} (f : x ⟶ y) (g : y ⟶ z) (hf : W f) (hg : multiplicativeClosure' W g) :
+    multiplicativeClosure' W (f ≫ g)
+
+variable {W : MorphismProperty C}
+
+/-- `multiplicativeClosure W` is multiplicative. -/
+instance : IsMultiplicative W.multiplicativeClosure where
+  id_mem x := .id x
+  comp_mem f g hf hg := by
+    induction hg with
+    | of _ hf₀ => exact .comp_of f _ hf hf₀
+    | id _ => rwa [Category.comp_id]
+    | comp_of f' g hf' hg h_rec =>
+      rw [← Category.assoc]
+      exact .comp_of (f ≫ f') g (h_rec f hf) hg
+
+/-- `multiplicativeClosure' W` is multiplicative. -/
+instance : IsMultiplicative W.multiplicativeClosure' where
+  id_mem x := .id x
+  comp_mem f g hf hg := by
+    induction hf with
+    | of _ h => exact .of_comp _ g h hg
+    | id _ => rwa [Category.id_comp]
+    | of_comp g' f hg' hf h_rec =>
+      rw [Category.assoc]
+      exact .of_comp g' (f ≫ g) hg' (h_rec g hg)
+
+/-- The multiplicative closure is greater or equal than the original property. -/
+lemma self_le_multiplicativeClosure : W ≤ W.multiplicativeClosure := fun {_ _} _ hf ↦ .of _ hf
+
+/-- The multiplicative closure of a multiplicative property is equal to itself. -/
+lemma eq_self_of_isMultiplicative [W.IsMultiplicative] : W.multiplicativeClosure = W := by
+  apply le_antisymm _ self_le_multiplicativeClosure
+  intro _ _ _ hf
+  induction hf with
+  | of _ hf₀ => exact hf₀
+  | id x => exact W.id_mem x
+  | comp_of _ _ _ hg hf => exact W.comp_mem _ _ hf hg
+
+/-- The multiplicative closure of `W` is the smallest multiplicative property greater or equal than
+`W`. -/
+lemma multiplicativeClosure_le_of_isMultiplicative_le
+    (W' : MorphismProperty C) [W'.IsMultiplicative]
+    (hWW' : W ≤ W') : W.multiplicativeClosure ≤ W' := by
+  intro _ _ _ hf
+  induction hf with
+  | of _ hf => exact hWW' _ hf
+  | id x => exact W'.id_mem _
+  | comp_of _ _ _ hg hf => exact W'.comp_mem _ _ hf (hWW' _ hg)
+
+lemma multiplicativeClosure_le_isMultiplicative (W' : MorphismProperty C) [W'.IsMultiplicative] :
+    multiplicativeClosure W ≤ W' ↔ W ≤ W' where
+  mp h := self_le_multiplicativeClosure.trans h
+  mpr h := multiplicativeClosure_le_of_isMultiplicative_le W' h
+
+lemma multiplicativeClosure_monotone (W' : MorphismProperty C) (hWW' : W ≤ W') :
+    W.multiplicativeClosure ≤ W'.multiplicativeClosure :=
+  multiplicativeClosure_le_of_isMultiplicative_le W'.multiplicativeClosure <|
+    hWW'.trans self_le_multiplicativeClosure
+
```

Comment:
It would be nice to add a lemma `multiplicativeClosure_eq_self_iff` saying ` W.multiplicativeClosure = W ↔ W.IsMultiplicative`.